ENH: Only apply first group once in fast GroupBy.apply

fjetter opened the pull request:
The issue of applying a function to the first row twice has been reported quite a few times (issues are usually referenced to #2936) and is also a documented shortcoming/implementation detail (see Notes in [docs](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.apply.html))

The argumentation is usually that this needs to be done to determine whether or not we can take a fast path for the calculation. Even if we're using the fast path for the calculation, however, the first row is still evaluated twice. I believe with this refactoring it would be possible to only eval the first row once iff the fast path is taken.
For functions with side effects or very expensive calculations in general this may be a big deal as discussed in various other issues.

I'm wondering if this small change may already help out folks. Truth be told, I'm not 100% certain about what's actually happening in this loop regarding the slider and item cache. For the use case where I encountered this, it seems to do the trick, though.

closes #2936
closes #2656
closes #7739
closes #10519
closes #12155
closes #20084
closes #21417

- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

fjetter committed:
ENH: Only apply first group once in fast GroupBy.apply

jreback reviewed (changes requested):
I am not sure you actually need to change, this rather you can modify code in pandas/core/groupby/generic.py and pandas/core/groupby/ops.py, search for ``fast_path``. 

but much more important is to actually have some tests for this.

fjetter commented:
I don't believe that changing the python code is actually sufficient since the Cython code I am modifying evaluates the first row twice, even if everything works as expected. My change here is essentially only using the calculation for the first row/test if the test is successful instead of recalculating it again.

I believe you are referring to https://github.com/pandas-dev/pandas/blob/453fa85a8b88ca22c7b878a3fcf97e068f11b6c4/pandas/core/groupby/ops.py#L174
where `fast_apply` is called and if successful, the result (calculated in Cython) is immediately returned. If an `InvalidApply` exception is caught, a slow, python-only path is taken. I'm not trying to optimize the slow/python path but I am arguing that even the `fast_apply` path will evaluate the first row twice although this isn't necessary.

Apart from implicit tests (this code section should be called at many places) there is one explicit test to ensure that the fast path can be taken and the flags for `mutated` are returned as expected, see https://github.com/pandas-dev/pandas/blob/453fa85a8b88ca22c7b878a3fcf97e068f11b6c4/pandas/tests/groupby/test_apply.py#L81
If there are other edge cases you can think of I can add more tests.

jreback reviewed (changes requested):
@fjetter this is not going to be acceptable w/o a test that fails before and passes now, meaning you have to show an explicit test that does this.

fjetter commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -1397,7 +1397,7 @@ def f(group):
     # we expect 2 zeros because we call ``f`` once to see if a faster route
     # can be used.
-    expected_names = [0, 0, 1, 2]
+    expected_names = [0, 1, 2]
```
@jreback This is the only test I am modifying in this PR. Essentially it tracks how often the function is applied to a given row/group by appending the groups name to a list. 
Previous behavior: Apply it twice to the first row. The first apply is done to check for `fast_apply` compatibility. The second apply is due to looping again from 0 to 2. 
In my change I propose to perform the initial `fast_apply` compatibility check _within_ the loop to save one iteration. 

Is this what you had in mind or what scenario would you like me to test?

jreback commented:
@fjetter I would like to collect all tests from any of the references issues. After you add them see if this closes them; list the issues in the top of the PR.

pep8speaks commented:
Hello @fjetter! Thanks for updating the PR.



 - In the file [`pandas/tests/groupby/test_apply.py`](https://github.com/pandas-dev/pandas/blob/0a075bd772ec4ba71c94eaf46071e04dae3aa615/pandas/tests/groupby/test_apply.py), following are the PEP8 issues :

> [Line 123:80](https://github.com/pandas-dev/pandas/blob/0a075bd772ec4ba71c94eaf46071e04dae3aa615/pandas/tests/groupby/test_apply.py#L123): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 79 characters)



 - In the file [`pandas/tests/series/test_apply.py`](https://github.com/pandas-dev/pandas/blob/0a075bd772ec4ba71c94eaf46071e04dae3aa615/pandas/tests/series/test_apply.py), following are the PEP8 issues :

> [Line 683:80](https://github.com/pandas-dev/pandas/blob/0a075bd772ec4ba71c94eaf46071e04dae3aa615/pandas/tests/series/test_apply.py#L683): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)

jreback commented on pandas/tests/frame/test_apply.py:
```diff
@@ -568,6 +568,37 @@ def test_apply_dup_names_multi_agg(self):
+        (1, [0, 1, 2, 3, 4, 5]),
+    ])
+    def test_apply_first_row_once(self, axis, expected):
+        df = pd.DataFrame({'a': [0, 0, 1, 1, 2, 2], 'b': np.arange(6)})
```
can you add the issue number as a comment

jreback commented on pandas/tests/frame/test_apply.py:
```diff
@@ -568,6 +568,37 @@ def test_apply_dup_names_multi_agg(self):
+            rows.append(row.name)
+            return 0
+        df.apply(f_fast, axis=axis)
+        # gh-2936
```
ahh you have here, well move to the top

jreback reviewed (changes requested):
pls add a comment with the issue number on the tests (if the test applies to multiple issue, then list both of them).

Add a whatsnew entry in 0.25.0. I would make this a sub-section in api-breaking changes, showing a mini-example and listing all of the closed issues.

jreback commented:
also we have a warning in the groupby.rst about this, can remove it.

fjetter commented:
I added a whatsnew entry and added the issue references to the tests. I kept the warning in `groupby.rst` since the double func evaluation may still occur if we need to fall back to the slow path.

jreback commented:
> I added a whatsnew entry and added the issue references to the tests. I kept the warning in groupby.rst since the double func evaluation may still occur if we need to fall back to the slow path.

that needs to be addressed here as well - i suspect this is actually the more common problem

fjetter commented:
I changed the implementation such that also the slow path evaluates the function only once. I refactored `apply_frame_axis0` to signal the sucess/validity of the fast apply via a status code instead of an exception. If the status code is 1, we fall back to the slow path calculation but will use the already calculated result for the first group.

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -1420,20 +1420,30 @@ def foo(x):
 
 def test_group_name_available_in_inference_pass():
     # gh-15062
+    # GH24748 ,GH2936, GH2656, GH7739, GH10519, GH12155, GH20084, GH21417
```
don’t add all the issues, just the ones that apply to each test; then it should be easy to see that we have all of the issues covered

fjetter replied:
I'll remove the links from this test since this test has been there before and is supposed to test whether the group name is available. The connection to this issue is merely coincidental.

For the (very similar) test in `groupby/test_apply.py` I added all these ticket numbers since you asked me to go through the issues and collect the tests. They are all duplicates and provide only very little additional information. All of them are linked to #2936 Is it ok if I remove all duplicates and only keep that one? same question for the whatsnew: list all duplicates or just the one reference issue?

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -26,6 +26,51 @@ Other Enhancements
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+(:issue:`2936`, :issue:`2656`, :issue:`7739`, :issue:`10519`, :issue:`12155`,
+:issue:`20084`, :issue:`21417`)
```
issues go at the end o the text

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -26,6 +26,51 @@ Other Enhancements
+
+.. code-block:: ipython
+
+    In [2]: df = pd.DataFrame({"a": ["x", "y"], "b": [1, 2]})
```
define df in a separate block above and show it

fjetter committed:
Address review comments

jreback commented:
i am ok with slightly duplicative tests if they are testing subtle behavior - if they are exact duplicates then ok to combine

fjetter commented:
The `test_group_name_available_in_inference_pass` is now a subset of `test_group_apply_once_per_group` in `groupby/test_apply.py`. Although the implementation is very similar I'd keep them both since they test different behavior

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -26,6 +26,44 @@ Other Enhancements
+    assert side_effects == ["x", "y"]
+
+
+(:issue:`2936`, :issue:`2656`, :issue:`7739`, :issue:`10519`, :issue:`12155`,
```
put the issue numbers at the end of the top paragraph.

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -26,6 +26,44 @@ Other Enhancements
+GroupBy.apply on ``DataFrame`` evaluates first group only once
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The implementation of ``DataFrame.groupby.apply`` previously evaluated func
```
see if you can make this a proper reference, I think:

``:meth:`DataFrameGroupBy.apply() <pandas.core.groupby.DataFrameGroupBy.apply>` ``

should work

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -26,6 +26,44 @@ Other Enhancements
+
+.. code-block:: ipython
+
+    In [6]: assert side_effects == ["x", "x", "y"]
```
just print side_effects (and it will show the 3 values)

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -26,6 +26,44 @@ Other Enhancements
+
+.. ipython:: python
+
+    assert side_effects == ["x", "y"]
```
same here, just show side_effects

jreback commented on pandas/_libs/reduction.pyx:
```diff
@@ -507,44 +507,41 @@ def apply_frame_axis0(object frame, object f, object names,
                 pass
 
             results.append(piece)
+            if status > 0:
```
this notion of status is pretty opaque, can you add some comments and/or make this much more obvious what you are doing.

jreback commented on pandas/tests/groupby/test_apply.py:
```diff
@@ -101,10 +101,34 @@ def f(g):
     splitter = grouper._get_splitter(g._selected_obj, axis=g.axis)
     group_keys = grouper._get_group_keys()
 
```
big comment on what this is testing and why this is the correct result

jreback commented on pandas/tests/groupby/test_apply.py:
```diff
@@ -101,10 +101,34 @@ def f(g):
 
 
+def test_group_apply_once_per_group():
+    # GH2936, GH7739, GH10519, GH2656, GH12155, GH20084, GH21417
```
add a 1-liner of what this is testing

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -26,6 +26,45 @@ Other Enhancements
+is safe to use a fast code path. Particularly for functions with side effects,
+this was an undesired behavior and may have led to surprises.
+
+Now every group is evaluated only a single time.
```
it is affter the sentence above.

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -26,6 +26,45 @@ Other Enhancements
+        return group
+    df.groupby("a").apply(func)
+
+Previous behavior:
```
use ``*Previous Behaviour*`` (not the ``*``) and similar on New

jreback commented on pandas/_libs/reduction.pyx:
```diff
@@ -507,44 +507,41 @@ def apply_frame_axis0(object frame, object f, object names,
+                piece = f(chunk)
+            except:
+                raise InvalidApply('Let this error raise above us')
+            # Need to infer if low level index slider will cause segfaults
```
blank line here

jreback commented on pandas/_libs/reduction.pyx:
```diff
@@ -507,44 +507,41 @@ def apply_frame_axis0(object frame, object f, object names,
                 pass
 
             results.append(piece)
+            if not successfull_fast_apply:
```
add a comment here on why you are returning early

jreback commented on pandas/core/groupby/ops.py:
```diff
@@ -165,26 +165,37 @@ def apply(self, f, data, axis=0):
         splitter = self._get_splitter(data, axis=axis)
         group_keys = self._get_group_keys()
-
+        reuse_result = False
```
this use of a flag is very odd and not very pythonic why is this needed at all?

jreback replied:
if you really really need this, then simply set ``result_values=None`` and test this

fjetter replied:
I've rewritten the logic to neither require the `status` identifier nor the `reuse_result` flag. The `reuse_result` flag was necessary in the previous implementation to catch the case for `MultiIndex` data which is handled by raising an exception in Cython. This could be handled differently if the multiindex check is performed outside of Cython but I didn't want to change too much about the current behavior. Having the index check close to the implementation seemed like a good idea to me

jreback commented on pandas/core/groupby/ops.py:
```diff
@@ -165,26 +165,37 @@ def apply(self, f, data, axis=0):
+                # Otherwise we need to fall back to the slow implementation.
+                if successful_fast_apply:
+                    return group_keys, fast_apply_result, mutated
+                else:
```
you don't need an else

jreback commented on pandas/tests/groupby/test_apply.py:
```diff
@@ -101,10 +101,43 @@ def f(g):
+    # GH2936, GH7739, GH10519, GH2656, GH12155, GH20084, GH21417
+
+    # This test should ensure that a function is only evaluted
+    # once per group. Previously the function has been evaluated twice
```
this really covers *all* of the above issues? meaning they all have exactly the same input? I find that hard to believe. Please parameterize as needed, but run thru the cases.

fjetter replied:
The test cases are in fact pretty close to each others. To make the review easier I included all test cases but I believe most are redundant. If you see no need for the duplication, I can easily remove / shrink the test again

fjetter committed:
More PR comments

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -26,6 +26,44 @@ Other Enhancements
+        return group
+    df.groupby("a").apply(func)
+
+Previous Behaviour:
```
can you add a ``*Previous Behavior:*`` (and same for New below)

jreback reviewed (changes requested):
very minor comment. pls merge master and ping on green.

jreback commented on pandas/core/groupby/ops.py:
```diff
@@ -165,26 +165,35 @@ def apply(self, f, data, axis=0):
         mutated = self.mutated
         splitter = self._get_splitter(data, axis=axis)
         group_keys = self._get_group_keys()
+        result_values = []
```
I think better to make this: ``result_values=None``

jreback commented on pandas/core/groupby/ops.py:
```diff
@@ -165,26 +165,35 @@ def apply(self, f, data, axis=0):
 
+            # If the fast apply failed we may still use the result
+            # for the first group
+            if len(result_values) == 1 and i == 0:
```
this would be then
``if result_values is None and i == 0``

fjetter committed:
PR comments

fjetter committed:
fix docs

WillAyd commented on pandas/tests/groupby/test_apply.py:
```diff
@@ -102,9 +102,89 @@ def f(g):
+@pytest.mark.parametrize(
+    "df,group_names",
+    [
+        (
```
Nitpicking but I find the readability here difficult. Can you start each set of parametrization with `(DataFrame(` and line break from there as required?

WillAyd commented on pandas/_libs/reduction.pyx:
```diff
@@ -507,40 +507,41 @@ def apply_frame_axis0(object frame, object f, object names,
+
+            # If the data was modified inplace we need to
+            # take the slow path to not risk segfaults
+            if require_slow_apply:
```
Might be overlooking but when would this ever be False?

jreback replied:
fixed up

WillAyd commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -63,6 +63,47 @@ is respected in indexing. (:issue:`24076`, :issue:`16785`)
+
+The implementation of :meth:`DataFrameGroupBy.apply() <pandas.core.groupby.DataFrameGroupBy.apply>`
+previously evaluated the supplied function consistently twice on the first group
+to infer if it is safe to use a fast code path. Particularly for functions with
```
Hmm I'm somewhat hesitant to market "side effects" as being fixed because I'm not sure we generally can make guarantees about mutations that occur during a groupby. Is it possible to leave this out and maybe use a relatively safe example like printing the groups instead?

jreback replied:
I agree with @WillAyd  here, showing the printing is enough (and a bit more clear).

jreback commented:
pls merge master

jreback commented:
can you merge master and fix the conflict

jreback commented:
can you merge master

jreback committed:
Merge branch 'master' into PR_TOOL_MERGE_PR_24748

jreback committed:
use printing in example

jreback commented:
rebased & whatsnew updated.

jreback committed:
fix whatsnew

jreback committed:
reformat tests

jreback committed:
small clean in reduction.pyx

jreback commented:
@WillAyd if you have any more comments.

WillAyd commented:
Thanks Jeff I’ll take a look tonight and merge if good

Sent from my iPhone

> On Mar 25, 2019, at 6:02 PM, Jeff Reback <notifications@github.com> wrote:
> 
> @WillAyd if you have any more comments.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into PERF/only_apply_first_row_once

WillAyd commented:
CI failure looked unrelated. Merged master will merge on green

WillAyd commented:
Thanks @fjetter !

CMCDragonkai commented:
Is this true only for groupby.apply? I tried test with just `apply` on a standard dataframe, and I found that the applied function is still being executed twice for the first row. This is quite problematic for me as my applied function is a side-effect.

fjetter commented:
@CMCDragonkai This issue is only fixed for the groupby.apply path. The ordinary `DataFrame.apply` uses different code paths where maybe a similar issue is buried.

CMCDragonkai commented:
Are you serious! So while groupby apply no longer has duplicate side effects, but normal apply still does!? Shouldn't it be easy to apply this fix to the situation for normal dataframe apply?

fjetter commented:
@CMCDragonkai I'd need to check what is causing the duplicate execution to say for sure. My gut tells me that it should be similarly possible. I can not guarantee it without actually checking it.

AlonMKovrr commented:
Are there any news regarding fixing this for the df.apply?

alonme commented:
Are there any news regarding fixing this for the df.apply?

MarcoGorelli commented:
> Are you serious! So while groupby apply no longer has duplicate side effects, but normal apply still does!?

@CMCDragonkai this is no way to address a community contributor who has volunteered their time to submit a pull request, please refrain from using such a tone in the future - thank you :)
